integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java

```diff
@@ -0,0 +1,633 @@
+import alluxio.security.authorization.Mode;
+import alluxio.util.ThreadUtils;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.file.InvalidPathException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Main FUSE implementation class.
+ * <p>
+ * Implements the FUSE callbacks defined by jni-fuse.
+ */
+@ThreadSafe
+public final class AlluxioJniFuseFileSystem extends AbstractFuseFileSystem {
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioJniFuseFileSystem.class);
+  private final FileSystem mFileSystem;
+  private final AlluxioConfiguration mConf;
+  // base path within Alluxio namespace that is used for FUSE operations
+  // For example, if alluxio-fuse is mounted in /mnt/alluxio and mAlluxioRootPath
+  // is /users/foo, then an operation on /mnt/alluxio/bar will be translated on
+  // an action on the URI alluxio://<master>:<port>/users/foo/bar
+  private final Path mAlluxioRootPath;
+  // Keeps a cache of the most recently translated paths from String to Alluxio URI
+  private final LoadingCache<String, AlluxioURI> mPathResolverCache;
+  private final LoadingCache<String, Long> mUidCache;
+  private final LoadingCache<String, Long> mGidCache;
+  private final AtomicLong mNextOpenFileId = new AtomicLong(0);
+  private final String mFsName;
+
+  private static final int LOCK_SIZE = 2048;
+  /** A readwrite lock pool to guard individual files based on striping. */
+  private final ReadWriteLock[] mFileLocks = new ReentrantReadWriteLock[LOCK_SIZE];
+
+  private final Map<Long, FileInStream> mOpenFileEntries = new ConcurrentHashMap<>();
+  private final Map<Long, FileOutStream> mCreateFileEntries = new ConcurrentHashMap<>();
+  private final boolean mIsUserGroupTranslation;
+
+  // To make test build
+  @VisibleForTesting
+  public static final long ID_NOT_SET_VALUE = -1;
+  @VisibleForTesting
+  public static final long ID_NOT_SET_VALUE_UNSIGNED = 4294967295L;
+  /**
+   * df command will treat -1 as an unknown value.
+   */
+  @VisibleForTesting
+  public static final int UNKNOWN_INODES = -1;
+  /**
+   * Most FileSystems on linux limit the length of file name beyond 255 characters.
+   */
+  @VisibleForTesting
+  public static final int MAX_NAME_LENGTH = 255;
+
+  private static final String USER_NAME = System.getProperty("user.name");
+  private static final String GROUP_NAME = System.getProperty("user.name");
```

Comment:
Is the group name also supposed to be the user name?

Reply:
it was like this all the time in our fuse integration.